### PR TITLE
security: Insert events for role change attempts

### PR DESCRIPTION
**Pull request by indradhanush**
This commit adds two new security event types:
- RoleChangeDenied
- RoleChangeGranted

At the moment, if a user sends an API request to change a user's role
to admin for example, while they are **already** an admin, we will
still end up recording an event for RoleChangeGranted (provided that
all checks pass and the write to the database is successful). This is
because we make no initial checks for current state before writing the
new expected state to the database. The fact that we want to add this
check or not remains outside the scope of this commit and subject to
further deliberation.

## Test Results

Following are the database entries in the `security_event_logs` table from using the `SiteUserIsSiteAdmin` API locally:

```
localhost sourcegraph@sourcegraph=# select * from security_event_logs order by id DESC limit 3;
┌─────┬───────────────────┬─────┬─────────┬───────────────────┬─────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────┬──────────────────────────────────┐
│ id  │       name        │ url │ user_id │ anonymous_user_id │ source  │                                                           argument                                                            │  version  │            timestamp             │
├─────┼───────────────────┼─────┼─────────┼───────────────────┼─────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────┼──────────────────────────────────┤
│ 129 │ RoleChangeDenied  │     │       1 │                   │ BACKEND │ {"by": 1, "to": "role_user", "for": 1, "from": "role_site_admin", "reason": "refusing to set current user site admin status"} │ 0.0.0+dev │ 2021-07-14 13:12:15.650317+05:30 │
│ 128 │ RoleChangeGranted │     │       1 │                   │ BACKEND │ {"by": 1, "to": "role_user", "for": 2, "from": "role_site_admin", "reason": ""}                                               │ 0.0.0+dev │ 2021-07-14 13:11:53.84756+05:30  │
│ 127 │ RoleChangeGranted │     │       1 │                   │ BACKEND │ {"by": 1, "to": "role_site_admin", "for": 2, "from": "role_user", "reason": ""}                                               │ 0.0.0+dev │ 2021-07-14 13:11:35.156653+05:30 │
└─────┴───────────────────┴─────┴─────────┴───────────────────┴─────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────┴──────────────────────────────────┘
(3 rows)

Time: 0.193 ms
```

And the local timestamp when this was tested:

```
$ date
Wed Jul 14 01:12:34 PM IST 2021
```

COREAPP-100
COREAPP-140

### CODE REVIEW

File: cmd/frontend/graphqlbackend/site_admin.go [hunk 1]
```diff
@@ -116,32 +120,77 @@ func (r *schemaResolver) DeleteOrganization(ctx context.Context, args *struct {
 	return &EmptyResponse{}, nil
 }
 
+type roleChangeEventArgs struct {
+	By     int32  `json:"by"`
+	For    int32  `json:"for"`
+	From   string `json:"from"`
+	To     string `json:"to"`
+	Reason string `json:"reason,omitempty"`
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/site_admin.go:128]: Let's include an empty reason. It's easier for the consumer to always receive all fields, even if some of them are blank

File: cmd/frontend/graphqlbackend/site_admin.go [hunk 2]
```diff
@@ -116,32 +120,77 @@ func (r *schemaResolver) DeleteOrganization(ctx context.Context, args *struct {
...
+	// At the moment, we log only two types of events:
+	// - RoleChangeDenied
+	// - RoleChangeGranted
+	//
+	// Unless we want to log another event for RoleChangeAttempted as well, invoking
+	// logRoleChangeAttempt before this point does not make sense since this is the first time in
+	// the lifetime of this function when we have all the details required for eventArgs, especially
+	// eventArgs.By which is used as the UserID in database.SecurityEvent - a required argument to
+	// write an entry into the database.
+	var eventName = database.SecurityEventNameRoleChangeDenied
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/site_admin.go:175]:
```suggestion
	eventName := database.SecurityEventNameRoleChangeDenied
```

File: cmd/frontend/graphqlbackend/site_admin.go [hunk 3]
```diff
@@ -116,32 +120,77 @@ func (r *schemaResolver) DeleteOrganization(ctx context.Context, args *struct {
...
+	if err = backend.CheckCurrentUserIsSiteAdmin(ctx, r.db); err != nil {
+		return nil, err
 	}
 
-	userID, err := UnmarshalUserID(args.UserID)
-	if err != nil {
+	if userResolver.ID() == args.UserID {
+		// err is passed as an argument to logRoleChangeAttempt. Returning errors.New directly will
+		// mean that this error is not logged in the reason for the security event.
+		err = errors.New("refusing to set current user site admin status")
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/site_admin.go:185]: You can get around this my naming the return arguments. In that case, returning the error directly will also assign it which allows it to be picked up by the defer.

Comment (ryanslade) [reply cmd/frontend/graphqlbackend/site_admin.go:185]: For example:

```go
func (r *schemaResolver) SetUserIsSiteAdmin(ctx context.Context, args *struct {
	UserID    graphql.ID
	SiteAdmin bool
}) (response *EmptyResponse, err error) {
 ...
}
```

**Review by ryanslade — APPROVED** [review]

File: cmd/frontend/graphqlbackend/site_admin.go (hunk 1 above)
Comment (indradhanush) [reply cmd/frontend/graphqlbackend/site_admin.go]: Makes sense! :100:

File: cmd/frontend/graphqlbackend/site_admin.go (hunk 3 above)
Comment (indradhanush) [reply cmd/frontend/graphqlbackend/site_admin.go]: Nice! I had not considered this approach! Thank you. This is much cleaner. :grinning:

**Comment by indradhanush** [comment]
For posterity's sake, here's the GQL query I used [locally](https://sourcegraph.test:3443/api/console?toast=integrations#%7B%22query%22%3A%22mutation%20SetUserIsSiteAdmin(%24userID%3A%20ID!%2C%20%24siteAdmin%3A%20Boolean!)%20%7B%5Cn%20%20setUserIsSiteAdmin(userID%3A%20%24userID%2C%20siteAdmin%3A%20%24siteAdmin)%20%7B%5Cn%20%20%20%20alwaysNil%5Cn%20%20%7D%5Cn%7D%5Cn%22%2C%22variables%22%3A%22%7B%5Cn%20%20%5C%22userID%5C%22%3A%20%5C%22VXNlcjox%5C%22%2C%5Cn%20%20%5C%22siteAdmin%5C%22%3A%20false%5Cn%7D%22%7D) for triggering the change.

```
mutation SetUserIsSiteAdmin($userID: ID!, $siteAdmin: Boolean!) {
  setUserIsSiteAdmin(userID: $userID, siteAdmin: $siteAdmin) {
    alwaysNil
  }
}
```

**Pull request merged**